### PR TITLE
Remove Result dependency

### LINKED ISSUE #145: Drop Result dependency and implement Swift 5 Result
[Swift 5 adds the `Result` type to the standary library.](https://github.com/apple/swift-evolution/blob/master/proposals/0235-add-result.md) It is no longer necessary to use a third-party dependency for `Result`. This dependency should be removed in favor the the native `Result` type.

### PR DESCRIPTION
**Pull request by daltonclaybrook**
Resolves #145

### CODE REVIEW

File: Sources/Commandant/Result+Additions.swift [hunk 1]
```diff
@@ -0,0 +1,27 @@
+//
+//  Result+Additions.swift
+//  Commandant
+//
+//  Created by Dalton Claybrook on 4/5/19.
+//  Copyright © 2019 Carthage. All rights reserved.
+//
+
+public extension Result {
```
Comment (mdiep) [inline Sources/Commandant/Result+Additions.swift]: These should not be public.

**Comment by ikesyo** [comment]
We should cut a release before merging this I think.

File: Sources/Commandant/Errors.swift [hunk 2]
```diff
@@ -33,6 +32,8 @@ extension CommandantError: CustomStringConvertible {
 	}
 }
 
+public enum NoError: Error {}
```
Comment (ikesyo) [inline Sources/Commandant/Errors.swift]: We don't need this, `Never` should be used instead.

File: Package@swift-5.swift [hunk 3]
```diff
@@ -0,0 +1,19 @@
+// swift-tools-version:5.0
+
+import PackageDescription
+
+let package = Package(
+    name: "Commandant",
+    products: [
+        .library(name: "Commandant", targets: ["Commandant"]),
```
[the remark this answers is not in the document]
Comment (ikesyo) [reply Package@swift-5.swift]: I'll add .swiftlint.yml and .hound.yml to exclude Package.swift from linting (in s separate PR).

File: Package@swift-5.swift [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    ],
+    dependencies: [
+        .package(url: "https://github.com/Quick/Quick.git", from: "2.0.0"),
+        .package(url: "https://github.com/Quick/Nimble.git", from: "8.0.0")
+    ],
+    targets: [
+        .target(name: "Commandant", dependencies: []),
+        .testTarget(name: "CommandantTests", dependencies: ["Commandant", "Quick", "Nimble"])
+    ],
+    swiftLanguageVersions: [.version("5")]
```
Comment (giginet) [inline Package@swift-5.swift]: you can use `.v5` https://github.com/apple/swift-package-manager/blob/dc27f4bc98324e54b7c95169c7afe1ed9455f59b/Sources/PackageDescription4/ManifestVersion.swift#L17

```suggestion
    swiftLanguageVersions: [.v5]
```

Comment (giginet) [reply Package@swift-5.swift]: https://github.com/apple/swift-package-manager/blob/dc27f4bc98324e54b7c95169c7afe1ed9455f59b/Sources/PackageModel/ToolsVersion.swift#L139

**Comment by daltonclaybrook** [comment]
Can anyone provide guidance on how best to resolve the Xcode 10.1 / Swift 4.2.1 failure? This is happening because the `SWIFT_VERSION` of 5.0 specified in the Xcode project file is unsupported in Xcode 10.1. Changing it back to 4.2 would resolve this issue, but I don't know if that makes much sense. The ideal scenario would be for Xcode to dynamically select the correct `SWIFT_VERSION` based on the latest available/supported Swift toolchain.

**Comment by ikesyo** [comment]
As I said [above](https://github.com/Carthage/Commandant/pull/146#issuecomment-480265245), by cutting a new release which supports Swift 4.2 before merging this, I think we don't need to keep Swift 4.2 compatibility in this PR.

**Comment by giginet** [comment]
An Ideal idea, all xcodeproj should be generated by its Package.swift. But don't be.
One of solutions, Separate build configuration on Xcode by each Swift versions.
Or you can drop Swift 4.2 support in next version.

**Comment by daltonclaybrook** [comment]
Thanks for the feedback @ikesyo & @giginet. If dropping Swift 4.2 support is the preferred approach, I will be happy to update this PR to do so.

**Comment by giginet** [comment]
swift_version on podspec should be 5.0
https://github.com/Carthage/Commandant/blob/master/Commandant.podspec#L31

And all Swift 4.2 should be removed from .travis.yml

**Comment by giginet** [comment]
You should also change `.swift-version`

https://github.com/Carthage/Commandant/blob/f40a9702d519c283d4fae73abdbb03a90d7cbddd/.swift-version

**Comment by giginet** [comment]
I noticed more issues.

Please remove dependency from podspec.
https://github.com/Carthage/Commandant/blob/a3c945494ba307a2df7695524ef988c5283f5a99/Commandant.podspec#L28

And you can remote `Cartfile` completely, because there are no public dependencies.

**Comment by giginet** [comment]
> Please remove dependency from podspec.

Sorry, already removed.

**Comment by giginet** [comment]
If project contains only `Cartfile.private`, it would work.
https://github.com/Carthage/Carthage/blob/master/Source/CarthageKit/Project.swift#L138-L139

So please remove empty `Cartfile`?
I'll approve this after removing.

File: Tests/CommandantTests/OptionsWithEnumProtocolSpec.swift [hunk 5]
```diff
@@ -124,15 +126,15 @@ struct TestEnumOptions: OptionsProtocol, Equatable {
...
+	typealias ClientError = Never
 	
 	static func create(_ a: StrictIntValue) -> (StrictStringValue) -> ([StrictStringValue]) -> ([StrictStringValue]?) -> (StrictStringValue?) -> (String) -> (StrictIntValue) -> ([String]) -> TestEnumOptions {
 		return { b in { c in { d in { e in { f in { g in { h in
 			return self.init(strictIntValue: a, strictStringValue: b, strictStringsArray: c, optionalStrictStringsArray: d, optionalStrictStringValue: e, optionalStrictInt: g, requiredName: f, arguments: h)
 			} } } } } } }
 	}
 	
-	static func evaluate(_ m: CommandMode) -> Result<TestEnumOptions, CommandantError<NoError>> {
+	static func evaluate(_ m: CommandMode) -> Result<TestEnumOptions, CommandantError<Never>> {
```
[the remark this answers is not in the document]
Comment (giginet) [reply Tests/CommandantTests/OptionsWithEnumProtocolSpec.swift]: This warning will be suppressed after merging #147

**Comment by jpsim** [comment]
> We should cut a release before merging this I think.

Agreed, projects like SourceKitten/SwiftLint aim to support Swift 4.1 through 5.0 (see https://github.com/jpsim/SourceKitten/pull/581), so we should have at least one release where those versions are supported. I opened #148 to release 0.16.0 before this is merged.

**Comment by giginet** [comment]
0.16.0 is already released. So I'll merge this 🎉

**Comment by ikesyo** [comment]
Looks like `Cartfile.resolved` was not updated.

**Pull request merged**